clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -88,18 +89,25 @@ abstract class IndexService[CI <: ClioIndex](
       }
   }
 
+  /**
+    * Update-or-insert (upsert) metadata for a given key.
+    *
+    * @param indexKey         The DTO for the key
+    * @param updatedMetadata  The DTO for the metadata.
+    * @return the ID for this upsert
+    */
   private def upsertToStream(
     indexKey: clioIndex.KeyType,
     updatedMetadata: clioIndex.MetadataType
   ) = {
-    Source.fromFuture(
-      persistenceService.upsertMetadata(
-        indexKey,
-        updatedMetadata,
-        documentConverter,
-        elasticsearchIndex
-      )
-    )
+    val document = documentConverter.document(indexKey, updatedMetadata)
+    val futureUpsertId = for {
+      _ <- persistenceDAO.writeUpdate(document, elasticsearchIndex)
+      _ <- searchDAO.updateMetadata(document)(elasticsearchIndex)
+    } yield {
+      ElasticsearchIndex.getUpsertId(document)
+    }
+    Source.fromFuture(futureUpsertId)
```

Comment:
Since the `Source.fromFuture` is just a convenience for the calling code, I think it'd be more clear to push the call there and have this method return a `Future`